### PR TITLE
Turn nested vectors into higher-dimensional arrays

### PR DESCRIPTION
**Pull request by tkoolen**
~~This passes all tests except the allocation ones. I'm working on something like your suggestion here: https://github.com/JuliaRobotics/LCMCore.jl/pull/26#discussion_r173220367 to fix those.~~

Ready for review now.

### CODE REVIEW

**Comment by tkoolen** [comment]
I'm confused. I made the changes to lift the array dimensions into the type system, but locally I'm still seeing allocations locally for `polynomial_t` and `polynomial_matrix_t` (but not `lcm_test_type_1` and `lcm_test_type_2`). That is already weird in itself, since if anything `polynomial_t` is simpler than `lcm_test_type_1`. But now the tests actually *do* pass on Travis...

@rdeits, could you check whether these tests pass on your machine(s)?

File: src/lcmtype.jl [hunk 1]
```diff
@@ -168,60 +182,55 @@ end
     exprs = Expr[]
```
Comment (tkoolen) [inline src/lcmtype.jl]: I'm no longer using `resize!` in `decodefield!`, but I think it may still be useful. Needs a separate test now though.

**Comment by tkoolen** [comment]
D'oh! The local failures were due to a version difference in BufferedStreams.jl. All good now.

File: src/lcmtype.jl (hunk 1 above)
Comment (tkoolen) [reply src/lcmtype.jl]: Done.

**Comment by tkoolen** [comment]
The `dimensions` change makes it so that this change is not backwards compatible, so I went ahead and removed the deprecated exports as well. Might want to release a version of LCMCore both before this PR and after.

**Comment by rdeits** [comment]
This looks great, thank you! Ok, I'll tag a patch release then merge this and tag a new minor version. Thanks!

**Pull request merged**